internal/cephfs/core/metadata.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	fsAdmin "github.com/ceph/go-ceph/cephfs/admin"
+)
+
+const (
+	// clusterNameKey cluster Key, set on RBD image.
+	clusterNameKey = "csi.ceph.com/cluster/name"
+)
+
+// ErrSubVolMetadataNotSupported is returned when set/get/list/remove subvolume metadata options are not supported.
+var ErrSubVolMetadataNotSupported = errors.New("subvolume metadata operations are not supported")
+
+func (s *subVolumeClient) supportsSubVolMetadata(err error) bool {
+	var invalid fsAdmin.NotImplementedError
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unsupported {
+		return false
+	} else if err != nil && errors.Is(err, &invalid) {
+		// In case the error is other than invalid command return error to the caller.
+		clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
+
+		return false
+	}
+	clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported
+
+	return true
+}
+
+// GetMetadata gets custom metadata on the subvolume in a volume based on key
+// name.
+func (s *subVolumeClient) GetMetadata(key string) (string, error) {
+	if !s.supportsSubVolMetadata(nil) {
```

Comment:
I dont think it's a better one. passing nil and err and making the decision on that one is not required; if this is required, have a new helper function to check the clusterAdditionalInfo and return the required error message.

Reply:
I just optimized it a bit, may be passing `nil` directly made you think something is not right, will be okay if I make it to:

```
func (s *subVolumeClient) supportsSubVolMetadata(err error) bool {
        var invalid fsAdmin.NotImplementedError
        if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unsupported {
                return false
        } else if err != nil && errors.Is(err, &invalid) {
                // In case the error is other than invalid command return error to the caller.
                clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported

                return false
        }
        clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported

        return true
}

// GetMetadata gets custom metadata on the subvolume in a volume based on key
// name.
func (s *subVolumeClient) GetMetadata(key string) (string, error) {
        var err error
        if !s.supportsSubVolMetadata(err) {
                return "", ErrSubVolMetadataNotSupported
        }
        fsa, err := s.conn.GetFSAdmin()
        if err != nil {
                return "", err
        }
        res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
        if !s.supportsSubVolMetadata(err) {
                return "", ErrSubVolMetadataNotSupported
        }
        if err != nil {
                return "", err
        }

        return res, nil
}
```

Reply:
I think I prefer two separate helpers to make their functioning clearer and easier to understand.

`s.isUnsupportedSubVolMetadata(err)`, pass the error, set either `supported` or `unsupported` and return 

`s.supportsSubVolMetadata()`, no error, just returns a bool. If the support was not checked before, it can do the checking if needed, or return `true` when in doubt. If it does checking, it should use `s.isUnsupportedSubVolMetadata(err)` to cache the support.